### PR TITLE
my php try refactoring

### PR DESCRIPTION
**Pull request by WatheqAlshowaiter**

### CODE REVIEW

**Comment by WatheqAlshowaiter** [comment]
@saleem-hadad

File: php/sms-template-detector/README.md [hunk 1]
```diff
@@ -14,4 +14,14 @@ The SMS template detector is a class that is responsible for parsing bank transa
 
 ```
 ./vendor/bin/pest
-```
\ No newline at end of file
+```
+
+## What I did
+- improved the test to cover more edge cases, e.g: using the other sms template
```
Comment (saleem-hadad) [inline php/sms-template-detector/README.md:20]: Nice one

File: php/sms-template-detector/README.md [hunk 2]
```diff
@@ -14,4 +14,14 @@ The SMS template detector is a class that is responsible for parsing bank transa
...
+- extract arrays to the class; they may be used in another way from the class
```
Comment (saleem-hadad) [inline php/sms-template-detector/README.md:21]: Good one, the functions should be mostly stateless unless holding data temporarily

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    private array $patterns = [
+        '/\{amount\}/' => '(.*?)',
+        '/\{brand\}/' => '(.*?)',
+        '/\{card\}/' => '(.*?)',
+    ];
+
+    public function detect(string $sms): ?array
+    {
+        foreach ($this->templates as  $template) {
+            $templateCopy = $template;
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php]: Do we need this line?

Comment (saleem-hadad) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php]: We can use the template directly in the replace function right?

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+            $templateCopy = preg_replace(
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php]: Maybe we can extract this into a separate private method

File: php/sms-template-detector/README.md [hunk 5]
```diff
@@ -14,4 +14,14 @@ The SMS template detector is a class that is responsible for parsing bank transa
...
+- change redundant `str_replace` to `foreach` based on the rules in the `$templates` & `$patterns`
```
Comment (saleem-hadad) [inline php/sms-template-detector/README.md:22]: Good one

File: php/sms-template-detector/v2/tests/SmsTemplateDetectorTest.php [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+use App\SmsTemplateDetector;
+
+test('it returns correct matched template with extracted data', function () {
+    $sut = new SmsTemplateDetector;
+
+    $smsTemplate = $sut->detect("Payment of 33.3 to Apple with 4233,");
+
+    $smsTemplate2 = $sut->detect("Purchase of 45.9 with 4853 at Samsung,");
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/tests/SmsTemplateDetectorTest.php:10]: Good one but it's always recommended to have one assertion per test per data
In case you have multiple dataset to test it's better to use data factory or in other languages they call it theory where you provide multiple data with the same assertion 
Something like this https://pestphp.com/docs/datasets

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                pattern: array_keys($this->patterns),
+                replacement: array_values($this->patterns),
+                subject: (string) $templateCopy
+            );
+
+
+            if (preg_match("/{$templateCopy}/", $sms, $matchedParts)) {
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php:27]: The condition itself can be extracted as a method something like 
if($this->isSmsMatchTemplate($templateCopy, $sms))

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+namespace App;
+
+class SmsTemplateDetector
+{
+    private array $templates = [
+        'Payment of {amount} to {brand} with {card},',
+        'Purchase of {amount} with {card} at {brand},',
+    ];
+
+    private array $patterns = [
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php:14]: Sine all the values are the same, Can you think of a better way to replace the placeholders {} without using this array ?

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        '/\{amount\}/' => '(.*?)',
+        '/\{brand\}/' => '(.*?)',
+        '/\{card\}/' => '(.*?)',
+    ];
+
+    public function detect(string $sms): ?array
+    {
+        foreach ($this->templates as  $template) {
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php:23]: Right now the for loop is doing multiple things like finding the template, replacing the placeholders and getting the sms parts. How can we make it more readable and doesn't do multiple things one one method

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php (hunk 4 above)
Comment (saleem-hadad) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php]: Method name can be something like 
$regexTemplate = $this->transormTemplateToRegex($template);

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        return null;
+    }
+
+    protected function getPartsWithValues(array $matchedParts, string $templateBody): array
+    {
+        $partsPositionsInTemplate = [];
+
+        if (strpos($templateBody, "{amount}")) {
```
Comment (saleem-hadad) [inline php/sms-template-detector/v2/app/SmsTemplateDetector.php:40]: How can we get rid of these multiple if conditions

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php (hunk 4 above)
Comment (WatheqAlshowaiter) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php]: yes, of course.. changed it

File: php/sms-template-detector/v2/tests/SmsTemplateDetectorTest.php (hunk 6 above)
Comment (WatheqAlshowaiter) [reply php/sms-template-detector/v2/tests/SmsTemplateDetectorTest.php:10]: something new to learn

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php (hunk 7 above)
Comment (WatheqAlshowaiter) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php:27]: how can I deal with `$matchedParts`. how can I return it if it is used in the line below?

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php (hunk 3 above)
Comment (WatheqAlshowaiter) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php]: yes we can :)

**Comment by WatheqAlshowaiter** [comment]
Thanks a lot @saleem-hadad  for your review. I really appreciate it

**Comment by saleem-hadad** [comment]
> Thanks a lot @saleem-hadad for your review. I really appreciate it

most welcome!

File: php/sms-template-detector/v2/app/SmsTemplateDetector.php (hunk 7 above)
Comment (saleem-hadad) [reply php/sms-template-detector/v2/app/SmsTemplateDetector.php:27]: you could split it into two functions 
isSmsMatchTemplate($templateCopy, $sms) will return bool
and another method let's call it something like 
```php
getSmsMatchedParts($templateCopy, $sms) {
    preg_match("/{$templateCopy}/", $sms, $matchedParts);
    return $matchedParts;
}
```
